### PR TITLE
libnvme/nvme-cli: add `nvme config-convert`, the legacy config to INI migration tool

### PR DESCRIPTION
**Pull request by martin-belanger**
This is the first thing that actually wires up to the INI read/write API added in #3566 and #3587: a one-shot, explicit command that converts an existing `config.json` and/or `discovery.conf` into the new INI format.

It's also just the first step. The plan is for a follow-on PR to add *implicit* conversion: `connect-all`/`discover` converting legacy config to INI automatically on first invocation, so INI becomes the format libnvme actually reads, with the conversion happening transparently at the CLI layer before any libnvme call. That implicit path is separate work, not part of this PR -- this one only adds the explicit command: `nvme config-convert`.

Commits:
- `nvme-cli: add the config.json reader for INI conversion`
- `fabrics, nvme-cli: add the discovery.conf reader for INI conversion`
- `nvme-cli: add the config-convert command`
- `libnvme/design: fix stale config-convert references in CONFIG.md`
- `doc: add nvme-config-convert man page`

`nvme config-convert` converts `config.json` and/or `discovery.conf` to the INI format read by `nvme connect-all`/`nvme discover`. With no `--config`, it converts whichever of the two well-known legacy files exist, matching what a bare `connect-all` already reads today; `--config` restricts it to one JSON file. Refuses to overwrite an existing target unless `--force` is given. On success, the converted legacy file(s) are renamed to `<name>.converted` so a repeated run doesn't convert them again.

Almost all of this lives in `config-convert.c`/`config-convert.h`, deliberately. Since support for the legacy configuration files is temporary, the conversion logic is kept isolated so it can eventually be removed with minimal impact. Unfortunately, not all the code could be put in `config-convert.c`. The only exception is the `discovery.conf` argv-style parser (`nvmf_convert_discovery_line`), which stays in `fabrics.c` to reuse the same option table as nvme discover; everything else lives in `config-convert.c`.

Ran the comments and man page through ChatGPT for a "man page style, simple/precise language" pass, then reviewed the result by hand.

Depends on #3587 (merged).

### CODE REVIEW

File: config-convert.c [hunk 1]
```diff
@@ -419,3 +422,127 @@ int nvme_config_convert_discovery(struct libnvmf_config_emitter *emitter,
...
+			converted_json = true;
+		}
+		if (!access(PATH_NVMF_DISC, F_OK)) {
+			ret = nvme_config_convert_discovery(emitter,
+							     PATH_NVMF_DISC);
+			if (ret)
+				goto out;
+			converted_disc = true;
+		}
+		if (!converted_json && !converted_disc) {
```
Comment (igaw) [inline config-convert.c]: this check is in the else block instead after the `if/else` block.

File: config-convert.c [hunk 2]
```diff
@@ -419,3 +422,127 @@ int nvme_config_convert_discovery(struct libnvmf_config_emitter *emitter,
...
+			nvme_show_error(
+				"nothing to convert: neither %s nor %s exists",
+				PATH_NVMF_CONFIG, PATH_NVMF_DISC);
+			ret = -ENOENT;
+			goto out;
+		}
+	}
```
Comment (igaw) [inline config-convert.c:561]: The documentation says 'one shot' migration, but in fact we have to call it twice, once for the json config and once for discovery.conf. I think it should do a one shot operation.

**Comment by igaw** [comment]
Looks good. Nice to see how simple the convert function is. Great job!

**Comment by igaw** [comment]
Could we also have some end to end tests for this feature (those in the `tests` directory?) We need to get this feature as correct as possible, including the corner cases if possible.

File: config-convert.c (hunk 2 above)
Comment (martin-belanger) [reply config-convert.c:561]: Good catch. I'll have that fixed.

File: config-convert.c (hunk 1 above)
Comment (martin-belanger) [reply config-convert.c]: Fixed

File: config-convert.c (hunk 2 above)
Comment (martin-belanger) [reply config-convert.c:561]: fixed

**Comment by martin-belanger** [comment]
Thanks for the review.

- Moved the "nothing to convert" check out of the `else` block, per your comment on config-convert.c:531. It fell out naturally from the fix below rather than needing a standalone change.
- Fixed the "one shot" issue (config-convert.c:538): discovery.conf conversion is no longer gated by `--config`. It matches `nvme connect-all`'s existing behavior now -- the well-known discovery.conf path is always converted when it exists, regardless of whether `--config` points at a non-default config.json. `--config` only ever affects which JSON file gets converted.
- Added end-to-end tests in `tests/nvme_config_convert_test.py`: DC/subsystem conversion, the legacy bare-array JSON format, dhchap-secret default inheritance and per-port override, and the error/force/rename paths.

One limitation on the test coverage worth flagging: discovery.conf's own conversion isn't exercised end-to-end. Its path has no `--config`-style override, so a test can only hit it by touching the real `/etc/nvme/discovery.conf` -- and that file exists on essentially every system, even freshly installed ones, since we ship an empty (comment-only) stub by default. Converting it end-to-end in a test would mean renaming that file on whatever machine runs the suite, which isn't something a test should do. The suite instead skips itself whenever that file is present, and I verified the logic separately against the real binary with the guard bypassed.

Related note, not for this PR: we should probably stop installing that empty discovery.conf stub by default (`meson.build` and `nvme.spec.in` both ship it unconditionally today). It's what made the test-coverage gap above universal rather than a corner case. Tracking as a separate follow-up.

**Comment by igaw** [comment]
Yes, we should remove the discovery.conf install step now. This should also be documented eventually. My current plan is to start documenting this for a release note file. All the stuff which could regress should be mentioned there.

**Comment by martin-belanger** [comment]
Sounds good — happy to help once you've got a structure in mind. What format are you picturing (a single `RELEASE_NOTES.md`, one entry per PR, something else)? I can contribute entries for the changes on my end (this discovery.conf removal, and whatever else from the INI/config pipeline is worth flagging) once there's a shape to fit them into.

File: tests/nvme_config_convert_test.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+    else 'nvme'
+
+_REAL_DISCOVERY_CONF = '/etc/nvme/discovery.conf'
+
+
+@unittest.skipIf(os.path.exists(_REAL_DISCOVERY_CONF),
+                  f'{_REAL_DISCOVERY_CONF} exists on this machine; '
+                  'config-convert always converts and renames it, so '
+                  'skipping to avoid touching real system configuration')
+class ConfigConvertCLITest(unittest.TestCase):
```
Comment (igaw) [inline tests/nvme_config_convert_test.py]: Is there a reason why this is not based on the TestNVMe class? It brings some helpers such run_cmd and all the logging infra.

File: tests/nvme_config_convert_test.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+        self.assertIn('nqn = nqn.2024-01.com.example:data.vol5', content)
+
+    def test_source_file_renamed_to_converted_on_success(self):
+        self._write_json({'hosts': []})
+        self._convert()
+        self.assertFalse(os.path.exists(self.config_json))
+        self.assertTrue(os.path.exists(self.config_json + '.converted'))
+
+
+if __name__ == '__main__':
```
Comment (igaw) [inline tests/nvme_config_convert_test.py:272]: I one test would also be if the config has been already be converted and the user calls 'nvme config-convert -J config.json' again. I think in this case config-convert should not error out instead doing nothing as there is already a config.json.converted,.e.g  the idempotent behavior.

File: tests/nvme_config_convert_test.py [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+import tempfile
+import unittest
+
+
+# Capture the nvme binary path from argv before unittest.main() strips it.
+_NVME_BIN = sys.argv[1] \
+    if len(sys.argv) > 1 and not sys.argv[1].startswith('-') \
+    else 'nvme'
+
+_REAL_DISCOVERY_CONF = '/etc/nvme/discovery.conf'
```
Comment (igaw) [inline tests/nvme_config_convert_test.py:42]: 60:#define PATH_NVMF_DISC		SYSCONFDIR "/nvme/discovery.conf"
294:	hfd->f = fopen(PATH_NVMF_DISC, "r");

we could use the new --set-options here:

`nvme --set-options discovery_conf=SOMEPATH/discovery.conf convert-config ...`

File: tests/nvme_config_convert_test.py (hunk 3 above)
Comment (martin-belanger) [reply tests/nvme_config_convert_test.py]: `TestNVMe.setUp()` unconditionally requires a real controller. None of that applies here. Subclassing it would mean overriding `setUp()` to skip essentially everything the base class does.

If `run_cmd`/the logging infra are the specific parts worth reusing, I could pull those out of `TestNVMe` into a small shared base class and use that. Let me know if that's worth doing, or if plain `subprocess.run()` (what's there now) is fine for a test this size.

File: tests/nvme_config_convert_test.py (hunk 4 above)
Comment (martin-belanger) [reply tests/nvme_config_convert_test.py:272]: Good idea. Will do.

File: tests/nvme_config_convert_test.py (hunk 5 above)
Comment (martin-belanger) [reply tests/nvme_config_convert_test.py:42]: `--set-options discovery_conf=...` needs the same plumbing a dedicated flag would: adding new param to the global ctx + setter/getter, etc. 

**There is a no-new-API alternative**: `unshare --user --map-root-user --mount` + a bind-mount over /etc/nvme, so the test's own discovery.conf shadows the real one ([man page](https://man7.org/linux/man-pages/man1/unshare.1.html)). No real root needed, but it depends on unprivileged user namespaces being allowed in whatever environment runs it -- not guaranteed everywhere, so it'd need a graceful skip where unavailable.

**Another option**: Since this only really needs to work in CI, and our CI containers are fully ours (ephemeral, thrown away after the job), we may not need either: the test can just back up /etc/nvme/discovery.conf, write known content, assert, and restore it afterward, gated behind an explicit CI-only env var. Locally it stays exactly as-is -- self-skips if a real discovery.conf exists. Simplest option: no new C code, no namespaces.

Comment (martin-belanger) [reply tests/nvme_config_convert_test.py:42]: Forget the `unshare` approach. This requires elevated privileges.

Comment (martin-belanger) [reply tests/nvme_config_convert_test.py:42]: Dug into where `tests/` actually runs before assuming any CI container is the right target. Turns out it doesn't run in the per-PR build matrix at all today -- `-Dnvme-tests=true` is only ever passed in `run-nightly-tests.yml`, which runs on a schedule against a privileged podman container (`ghcr.io/linux-nvme/debian:latest`), not on every PR push. `meson install` doesn't run there either, so I pulled that image (and the fedora one) directly to check: neither has nvme-cli installed, and /etc/nvme doesn't even exist. So discovery.conf genuinely isn't present in that container unless we put it there ourselves.

Given that, the env-var-gated backup/write/restore approach from above should work cleanly there: no nvme-cli package to conflict with, a privileged/root container, thrown away after each nightly run. I'd suggest we only wire this into the nightly workflow for now rather than try to get it into the per-PR matrix -- nightly-only discovery.conf coverage seems like a reasonable bar given the constraints, and it's already the same tier `registry-cli` gets today (that one doesn't run per-PR either, same infrastructure gap, unrelated to this PR).

What do you think?

Comment (igaw) [reply tests/nvme_config_convert_test.py:42]: Hmm, I think it's not worth the effort for discovery.conf. As far I can tell we can cover all scenarios with config.json and the only gain with discovery.conf is the parsing bits, which is already covered in libnvme as far I remember

Let's go without it I'd say.

File: tests/nvme_config_convert_test.py (hunk 3 above)
Comment (igaw) [reply tests/nvme_config_convert_test.py]: Ah, didn't realize that. I would like to use the same infra for command logging we have currently. So yeah, best idea is to move the common code in a base class I'd say.

Comment (martin-belanger) [reply tests/nvme_config_convert_test.py]: Will do.

**Comment by igaw** [comment]
> Sounds good — happy to help once you've got a structure in mind. What format are you picturing (a single `RELEASE_NOTES.md`, one entry per PR, something else)? I can contribute entries for the changes on my end (this discovery.conf removal, and whatever else from the INI/config pipeline is worth flagging) once there's a shape to fit them into.

I don't have an exact idea how it should be handled, I think when a new feature or so lands. I'd say we just copy the strategy/format from an existing project, systemd? 

What about RELEASE.md

Anyway w oould make my life simpler :)

File: tests/nvme_config_convert_test.py (hunk 3 above)
Comment (martin-belanger) [reply tests/nvme_config_convert_test.py]: Done

File: tests/nvme_config_convert_test.py (hunk 4 above)
Comment (martin-belanger) [reply tests/nvme_config_convert_test.py:272]: Done

**Comment by martin-belanger** [comment]
All done. 
I will have a closer look at what systemd does when they have a new relase. From what I can tell they use a "NEWS" file.

**Comment by igaw** [comment]
Okay, let's go with this and see how it goes. Thanks a lot!
